[sonos] Firmware update for Ikea Symfonisk speakers breaking discovery in OpenHAB sonos binding

It looks like, that some time in the last 24 hours, an update was rolled out to the Ikea Symfonisk speakers.

My bookshelf speakers were back in my inbox this morning (both in OH 2.5.12 and OH 3.1.0M2), named:
SYMFONISK Bookshelf (Dining Room) and SYMFONISK Bookshelf (Office). 2 of each.

The old properties content (before deleting the device from my Things) was:
hardwareVersion	1.20.3.3-2.0
identification	Office
ipAddress	10.55.199.122
macAddress	34:7E:5C:F8:05:9C
modelId	SYMFONISK
serialNumber	34-7E-5C-F8-05-9C:8
vendor	IKEA

The new properties content (after adding the duplicated detected device) is:
hardwareVersion | 1.20.3.3-2.0
identification | Office
ipAddress | 10.55.199.122
macAddress | 34:7E:5C:F8:05:9C
modelId | ZonePlayer
serialNumber | 34-7E-5C-F8-05-9C:8
vendor | SONOS

So the reason for the duplication is, that the modelId has changed and hence, so that the identifier for the speaker has changed from sonos:SYMFONISK:RINCON_347E5CF8059C01400 to sonos:zoneplayer:RINCON_347E5CF8059C01400

That's just a minor matter. It seems, that also the naming of the discovered item has changed.

The original item "Represents IKEA SYMFONISK speaker".

The new item "The Zone Player represents a Sonos Zone Player which is not known to this binding" and comes up in the discovery as: "SYMFONISK Bookshelf (Dining Room)". It's been a long time, since I added these speakers in my OH2.5, so I can't recall, what they were called back then, but I believe it was indeed "IKEA SYMFONISK speaker (Dining Room)" or something the likes.

Now, here's where things go bad: while the bookshelf speakers are discovered, the table lamp speakers are not. They would now probably be discovered as "SYMFONISK Table lamp (Room)".

OH 3.1.0M2 is throwing the following error in openhab.log:
2021-03-12 15:25:51.946 [ERROR] [y.upnp.internal.UpnpDiscoveryService] - Participant 'org.openhab.binding.sonos.internal.discovery.ZonePlayerDiscoveryParticipant' threw an exception
java.lang.IllegalArgumentException: ID segment 'Table lamp' contains invalid characters.

I tried to increasing logging to debug level for the Sonos binding, but it was not more conclusive. Clearly introducing the space in the device type/name has broken the strict Thing naming of OpenHAB. And by the looks both 2.5.12 and 3.1.0M2, as neither of them have rediscovered my table lamps.

/M

--- a/bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/discovery/ZonePlayerDiscoveryParticipant.java
+++ b/bundles/org.openhab.binding.sonos/src/main/java/org/openhab/binding/sonos/internal/discovery/ZonePlayerDiscoveryParticipant.java
@@ -117,7 +117,10 @@ public class ZonePlayerDiscoveryParticipant implements UpnpDiscoveryParticipant 
     }
 
     private String getModelName(RemoteDevice device) {
-        return SonosXMLParser.extractModelName(device.getDetails().getModelDetails().getModelName());
+        // For Ikea SYMFONISK models, the model name now starts with "SYMFONISK" with recent firmwares
+        // We can no more use extractModelName as it deletes the first word ("Sonos" for all other devices)
+        return device.getDetails().getModelDetails().getModelName().toUpperCase().contains("SYMFONISK") ? "SYMFONISK"
+                : SonosXMLParser.extractModelName(device.getDetails().getModelDetails().getModelName());
     }
 
     private @Nullable String getSonosRoomName(RemoteDevice device) {

